distributed/shuffle/tests/test_shuffle.py

```diff
@@ -1481,8 +1354,274 @@ async def receive(self, data: list[bytes]) -> None:
+        self.block_shuffle_receive = asyncio.Event()
+
+    async def shuffle_receive(self, *args: Any, **kwargs: Any) -> None:
+        self.in_shuffle_receive.set()
+        await self.block_shuffle_receive.wait()
+        return await super().shuffle_receive(*args, **kwargs)
+
+
+@mock.patch.dict(
+    DEFAULT_EXTENSIONS,
+    {"shuffle": BlockedShuffleReceiveShuffleWorkerExtension},
+)
+@gen_cluster(client=True, nthreads=[("", 1)] * 2)
+async def test_deduplicate_stale_transfer(c, s, a, b):
+    df = dask.datasets.timeseries(
+        start="2000-01-01",
+        end="2000-01-10",
+        dtypes={"x": float, "y": float},
+        freq="100 s",
+    )
+    out = dd.shuffle.shuffle(df, "x", shuffle="p2p")
+    out = out.persist()
+
+    shuffle_extA = a.extensions["shuffle"]
+    shuffle_extB = b.extensions["shuffle"]
+    await asyncio.gather(
+        shuffle_extA.in_shuffle_receive.wait(), shuffle_extB.in_shuffle_receive.wait()
+    )
+    del out
+
+    while s.tasks or shuffle_extA.shuffles or shuffle_extB.shuffles:
+        await asyncio.sleep(0)
+
+    out = dd.shuffle.shuffle(df, "x", shuffle="p2p")
+    x, y = c.compute([df.x.size, out.x.size])
+    await get_shuffle_id(s)
+    shuffle_extA.block_shuffle_receive.set()
+    shuffle_extB.block_shuffle_receive.set()
+
+    x = await x
+    y = await y
+    assert x == y
+
+    await clean_worker(a, timeout=2)
+    await clean_worker(b, timeout=2)
+    await clean_scheduler(s, timeout=2)
+
+
+class BlockedBarrierShuffleWorkerExtension(ShuffleWorkerExtension):
+    def __init__(self, worker: Worker) -> None:
+        super().__init__(worker)
+        self.in_barrier = asyncio.Event()
+        self.block_barrier = asyncio.Event()
+
+    async def _barrier(self, *args: Any, **kwargs: Any) -> int:
+        self.in_barrier.set()
+        await self.block_barrier.wait()
+        return await super()._barrier(*args, **kwargs)
+
+
+@mock.patch.dict(
+    DEFAULT_EXTENSIONS,
+    {"shuffle": BlockedBarrierShuffleWorkerExtension},
+)
+@gen_cluster(client=True, nthreads=[("", 1)] * 2)
+async def test_handle_stale_barrier(c, s, a, b):
+    df = dask.datasets.timeseries(
+        start="2000-01-01",
+        end="2000-01-10",
+        dtypes={"x": float, "y": float},
+        freq="100 s",
+    )
+    out = dd.shuffle.shuffle(df, "x", shuffle="p2p")
+    out = out.persist()
+
+    shuffle_extA = a.extensions["shuffle"]
+    shuffle_extB = b.extensions["shuffle"]
+
+    wait_for_barrier_on_A_task = asyncio.create_task(shuffle_extA.in_barrier.wait())
+    wait_for_barrier_on_B_task = asyncio.create_task(shuffle_extB.in_barrier.wait())
+
+    await asyncio.wait(
+        [wait_for_barrier_on_A_task, wait_for_barrier_on_B_task],
+        return_when=asyncio.FIRST_COMPLETED,
+    )
+    del out
+
+    while s.tasks:
+        await asyncio.sleep(0)
+
+    out = dd.shuffle.shuffle(df, "x", shuffle="p2p")
+    x, y = c.compute([df.x.size, out.x.size])
+    await get_shuffle_id(s)
```

Comment:
what is this doing/asserting? This looks like there is an implicit assumption about `get_shuffle_id` doing something. ideally we'd assert on some sort of state here. IF that's not possible, I suggest to add comment to explain what this is for

Reply:
I've renamed it. This waits until the shuffle is initialized on the scheduler and returns its ID.